include/wlr/types/wlr_buffer.h

```diff
@@ -21,23 +21,51 @@ struct wlr_buffer_impl {
 		struct wlr_dmabuf_attributes *attribs);
 };
 
+/**
+ * A buffer containing pixel data.
+ *
+ * A buffer has a single producer (the party who created the buffer) and
+ * multiple consumers (parties reading the buffer). When all consumers are done
+ * with the buffer, it gets released and can be re-used by the producer. When
+ * the producer and all consumers are done with the buffer, it gets destroyed.
+ */
 struct wlr_buffer {
 	const struct wlr_buffer_impl *impl;
 
-	size_t n_refs;
+	int width, height;
+
+	bool dropped;
+	size_t n_locks;
+
+	struct {
+		struct wl_signal destroy;
+		struct wl_signal release;
+	} events;
 };
 
+/**
+ * Initialize a buffer. This function should be called by producers. The
+ * initialized buffer is referenced: once the producer is done with the buffer
+ * they should call wlr_buffer_drop.
+ */
 void wlr_buffer_init(struct wlr_buffer *buffer,
-	const struct wlr_buffer_impl *impl);
+	const struct wlr_buffer_impl *impl, int width, int height);
 /**
- * Reference the buffer.
+ * Unreference the buffer. This function should be called by producers when
+ * they are done with the buffer.
  */
-struct wlr_buffer *wlr_buffer_ref(struct wlr_buffer *buffer);
+void wlr_buffer_drop(struct wlr_buffer *buffer);
```

Comment:
Is this function really necessary? There are only 2 potential users of this function, which is wlr_client_buffer, and a swapchain.
wlr_client_buffer doesn't need it, because it can just destroy the buffer on the release signal.
The swapchain can just implement this itself, taking the release event into account.

Reply:
Do you mean that we can replace `wlr_buffer_drop` with `wlr_buffer_destroy`?

This is necessary when e.g. resizing a swapchain: the old buffers (with the old size) can be dropped by the swapchain while still being used by e.g. DRM for scanout.

Reply:
And in `wlr_surface`, we'd need to have a list of buffers the compositor is still using and a bunch of release listeners to only destroy buffers when we can.

Reply:
>This is necessary when e.g. resizing a swapchain: the old buffers (with the old size) can be dropped by the swapchain while still being used by e.g. DRM for scanout.

My point is that you'd still keep roughly the same logic, but it would be implement in the swapchain rather than in `wlr_buffer`.

>And in wlr_surface, we'd need to have a list of buffers the compositor is still using and a bunch of release listeners to only destroy buffers when we can.

Isn't that just done by the reference counting?

Reply:
>My point is that you'd still keep roughly the same logic, but it would be implement in the swapchain rather than in wlr_buffer.

If the swapchain has 3 slots and is resized, there are 3 "dangling buffers" still used by DRM that will need to be destroyed when released.

>Isn't that just done by the reference counting?

What do you suggest actually? I realize I may have misunderstood.

Do you want to remove `wlr_buffer_drop` and replace it with a `wlr_buffer_destroy` function which just immediately destroys the buffer? If that's the case, someone needs to keep track of the release event and destroy the buffer when it happens. I don't want to duplicate this logic in `wlr_surface` and in the swapchain.